### PR TITLE
Add StyleTTS 2

### PR DESCRIPTION
**Pull request by eustlb**
# What does this PR do?

Adds [StyleTTS 2](https://huggingface.co/papers/2306.07691) to support the original model but also other checkpoints like [Kokoro](https://huggingface.co/hexgrad/Kokoro-82M).

🆕 🔥 **This implementation also add batch support (early benchmarks show ~50% inference speed improvement for BS 128) and mask support for padded inputs.** 

> [!NOTE]  
> This implementation differs slightly from [the original codebase](https://github.com/yl4579/StyleTTS2). The aim here is to have the clearest possible correspondence of naming and structure compared to the original papers ([StyleTTS 2](https://arxiv.org/abs/2306.07691) that builds on top of [StyleTTS](https://arxiv.org/abs/2205.15439)).

# TODO
- [x] modeling code →  **ready for review**
- [x] tokenization →  add other languages + docstring
- [ ] processor →  add docstring
- [ ] tests
- [ ] model_doc

# Benchmarks

## BS 1 
- **~8%** faster than the original codebase (benchmarking script: [original](https://gist.github.com/eustlb/5a9092a1635c967dfb908cc2ac757780#file-benchmark_original-py) and [transformers](https://gist.github.com/eustlb/5a9092a1635c967dfb908cc2ac757780#file-benchmark_trfms_forward-py))

### CODE REVIEW

**Comment by eustlb** [comment]
cc @Cyrilvallez, modeling code is ready for review 🤗

**Review by Cyrilvallez — COMMENTED** [review]
Alright! Super sorry about the (very) long delay! Very very nice implementation, not much to complain about here! Great work! 🤗
Mostly just a few recurrent but easy to fix stuff:
- Use only the config in `__init__` if possible. I did not check each one, and some may have additional params that do not really make sense in the config as they are hard-coded, such as the boolean args in `StyleTextToSpeech2AdainResBlock1d`, but let's still try to use config as much as possible (and add additional parameters if still needed for internal-only purposes)
- If possible, all the `transpose` ops should have explicit dim number, let's try not to use `-1` if number of dims is always the same (not sure here)
- The weight_norm function should probably always be one or the other, as layer names depend on it (if needed, let's use a hard pytorch version check for this one)

**Comment by Cyrilvallez** [comment]
Also, I assume modular could not be used at all here?

File: src/transformers/models/style_text_to_speech_2/modeling_style_text_to_speech_2.py [hunk 1]
```diff
@@ -0,0 +1,1207 @@
...
+            padding=config.kernel_size // 2,
+        )
+        self.norm = nn.LayerNorm(config.hidden_size)
+        self.leaky_relu_slope = config.leaky_relu_slope
+        self.dropout = nn.Dropout(config.dropout)
+
+        # apply weight norm
+        weight_norm = nn.utils.weight_norm
+        if hasattr(nn.utils.parametrizations, "weight_norm"):
+            weight_norm = nn.utils.parametrizations.weight_norm
```
[the remark this answers is not in the document]
Comment (eustlb) [reply src/transformers/models/style_text_to_speech_2/modeling_style_text_to_speech_2.py:150]: True! We do not need to check for older torch versions right? since `parametrizations.weight_norm` is supported anyway for torch>=2.1

File: src/transformers/models/style_text_to_speech_2/configuration_style_text_to_speech_2.py [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+        Instantiate a [`StyleTextToSpeech2Config`] (or a derived class) from StyleTextToSpeech2 sub-models configuration.
+
+        Returns:
+            [`StyleTextToSpeech2Config`]: An instance of a configuration object
+        """
+        return cls(
+            acoustic_text_encoder_config=acoustic_text_encoder_config.to_dict(),
+            predictor_config=predictor_config.to_dict(),
+            decoder_config=decoder_config.to_dict(),
+            **kwargs,
```
[the remark this answers is not in the document]
Comment (eustlb) [reply src/transformers/models/style_text_to_speech_2/configuration_style_text_to_speech_2.py]: Indeed, not really necessary 😊 changed the implem anyway for something that is more aligned with what we do for other models with subconfis

File: src/transformers/models/style_text_to_speech_2/modeling_style_text_to_speech_2.py [hunk 3]
```diff
@@ -0,0 +1,1207 @@
...
+        self.bert_model = AutoModel.from_config(config.prosodic_text_encoder_sub_model_config)
+        self.proj_out = nn.Linear(
+            config.prosodic_text_encoder_sub_model_config.hidden_size, 
+            config.hidden_size
+        )
+
+    def forward(self, input_ids, input_lengths=None):
+        batch_size, seq_len = input_ids.shape
+        if input_lengths is None:
+            attention_mask = torch.ones_like(input_ids).to(input_ids.device)
```
[the remark this answers is not in the document]
Comment (eustlb) [reply src/transformers/models/style_text_to_speech_2/modeling_style_text_to_speech_2.py]: Nice catch !! 😅

File: src/transformers/models/style_text_to_speech_2/modeling_style_text_to_speech_2.py [hunk 4]
```diff
@@ -0,0 +1,1207 @@
...
+        hidden_states = self.conv2(hidden_states.transpose(1, -1)).transpose(1, -1)
+        hidden_states = _mask_hidden_states(hidden_states, input_lengths)
+
+        return hidden_states
+
+    def _shortcut(self, hidden_states):
+        hidden_states = self.upsample(hidden_states.transpose(1, -1))
+        hidden_states = self.conv1_shortcut(hidden_states).transpose(1, -1)
+
+        return hidden_states
```
[the remark this answers is not in the document]
Comment (eustlb) [reply src/transformers/models/style_text_to_speech_2/modeling_style_text_to_speech_2.py]: True! thanks, it's indeed clearer this way

File: src/transformers/models/style_text_to_speech_2/modeling_style_text_to_speech_2.py [hunk 5]
```diff
@@ -0,0 +1,1207 @@
...
+    def forward(self, hidden_states, style, input_lengths):
+        x = self.norm1(hidden_states, style, input_lengths)
+        x = x + (1 / self.alpha1) * (torch.sin(self.alpha1 * x) ** 2)
+        x = self.conv1(x.transpose(1, -1)).transpose(1, -1)
+        x = _mask_hidden_states(x, input_lengths)
+
+        x = self.norm2(x, style, input_lengths)
+        x = x + (1 / self.alpha2) * (torch.sin(self.alpha2 * x) ** 2)
+        x = self.conv2(x.transpose(1, -1)).transpose(1, -1)
+        x = _mask_hidden_states(x, input_lengths)
```
[the remark this answers is not in the document]
Comment (eustlb) [reply src/transformers/models/style_text_to_speech_2/modeling_style_text_to_speech_2.py]: yep totally agree here

**Comment by monuminu** [comment]
any update on this ?

**Comment by eustlb** [comment]
Hey @monuminu, this PR went stale due to other priorities. I’ll try to get it moving ASAP, though I’m not sure I’ll manage before the summer break. Thanks for checking in, I’ll bump its priority.

**Comment by yao-matrix** [comment]
@eustlb , do you still plan to enable this model architecture? Thx very much.

**Comment by rhulha** [comment]
I hope Kokoro has earned the energy for someone smarter than me to merge this PR.

Kokoro has been such a great TTS library and project.

Kokoro will go down in history as the first, free, high quality TTS lib that spread to the masses and helped so many many developers:

Shameless plug: https://github.com/rhulha/StreamingKokoroJS

Let's honor that brave little TTS and make it work in the bravest and best AI library in the universe ! This one ! 

It would be a shame to not have support for such a historic TTS lib.

Please please pretty please!
